File: LeetCode/1593. Split a String Into the Max Number of Unique Substrings/solution.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        n = len(s)
        ans = 0
        def dfs(i, path, word):
            nonlocal ans
            if i == n:
                if word not in path:
                    ans = max(ans, len(path)+(word != ''))
                return
            word += s[i]
            if word not in path:
                path.add(word)
                dfs(i+1, path, '')
                path.remove(word)
            dfs(i+1, path, word)
            return 
        dfs(0,set(),'')
        return ans 



class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        n = len(s)
        res = [0]
        
        def dfs(idx, path):
            if idx == n:
                if len(set(path)) == len(path):
                    res[0] = max(res[0], len(path))
                return
            if path:
                dfs(idx+1, path[:-1]+[path[-1]+s[idx]])
            dfs(idx+1, path+[s[idx]])
            return

        
        dfs(0, [])
        return res[0]

File: LeetCode/1593. Split a String Into the Max Number of Unique Substrings/solution.cpp
class Solution {
public:
    int maxUniqueSplit(string s) {
        unordered_set<string> visited;
        int n = s.length();
        function<int(int)> dfs = [&](int cur){
            if (cur == n)
                return 0;
            int res = -1;
            for (int nxt = cur+1; nxt <= n; nxt++){
                string word = s.substr(cur,nxt-cur);
                if (1+n-nxt <= res)
                    break;
                if (visited.count(word) == 0){
                    visited.insert(word);
                    int res_ = dfs(nxt);
                    if (res_ > -1)
                        res = max(res, 1+res_);
                    visited.erase(word);
                }
            }
            return res;
        };
        return dfs(0);
    }
};